Fixes #26801 - implements orderable multi select

ezr-ondrej opened the pull request:
*Updated for typeahead*

![draggable_typeahead](https://user-images.githubusercontent.com/2884324/57875532-78ecb000-7813-11e9-8e66-b427be72c6cf.gif)
Enhance typeahead using **react-dnd** powerful drag&drop engine, we could leverage more.

ezr-ondrej committed:
Fixes #26801 - implements orderable multi select

Ron-Lavi commented:
I really like it, looks awesome !
but Patternfly has the Typeahead Select implementation which got approved by their designers,
I think we should align with them and use similar components over our app for searching, autocomplete, etc.. What I like about the implementation there is the fact that the results are being highlighted when you filter them, and the autocomplition where you can see the result in sort of a placeholder.

and what I like in your PR is the dnd feature which maybe we can contribute to `react-bootstrap-typeahead`
 
[Typeahead Select in PF](https://patternfly-react.surge.sh/patternfly-3/index.html?knob-Clear%20Button=true&knob-Multiple%20Selection=true&knob-highlight%20Only%20Result=true&knob-select%20Hint%20On%20Enter=true&selectedKind=patternfly-react%2FForms%20and%20Controls%2FTypeAhead%20Select&selectedStory=Multiple%20Selection&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)
![patternfly-typeahead-multiple](https://user-images.githubusercontent.com/26363699/57842384-0ceb5700-77d5-11e9-91c7-05b10d3455f9.gif)

ezr-ondrej commented:
Ah, didn't look properly :man_facepalming: The main thing is the dnd though and it is pretty easy to reimplement for typeahead.

ezr-ondrej commented:
Rewritten for the typeahead:
![draggable_typeahead](https://user-images.githubusercontent.com/2884324/57875532-78ecb000-7813-11e9-8e66-b427be72c6cf.gif)

ares commented:
could we get review from @theforeman/ui-ux please?

xprazak2 commented on webpack/assets/javascripts/react_app/components/common/forms/OrderableSelect/OrderableSelect.js:
```diff
@@ -0,0 +1,62 @@
+  ...props
+}) => {
+  const [internalValue, setInternalValue] = useState(
+    (value || defaultValue || [])
```
should `defaultValue` be `[]`?

xprazak2 commented on webpack/assets/javascripts/react_app/components/common/forms/OrderableSelect/components/OrderableToken.js:
```diff
@@ -0,0 +1,44 @@
+  tabIndex,
+  labelKey,
+}) => (
+  <Token disabled={disabled} onRemove={onRemove} tabIndex={tabIndex}>
```
I think we can get Token from patternfly-react, maybe something like:

```js
import { TypeAheadSelect } from 'patternfly-react';
// snip
<TypeAheadSelect.Token>
```

ezr-ondrej replied:
~~We can't because patternfly-react is not exporting it :(~~
*Edit*: My bad, it does, you are right, happy to do so! :)

xprazak2 commented:
Works for me, left a couple of comments.

ezr-ondrej commented on package.json:
```diff
@@ -98,8 +99,11 @@
     "prop-types": "^15.6.0",
     "react": "^16.8.1",
     "react-ace": "^6.3.2",
+    "react-bootstrap-typeahead": "^3.4.3",
```
no idea why this one is not pulled as a dependency, do not want to state it manually now :(
Ideas?

xprazak2 replied:
It got installed into node_modules for me when I did the following:

1) I removed it from package.json
2) rm -rf node_modules
3) npm install

ezr-ondrej replied:
that worked :+1:

ezr-ondrej commented:
Thanks for review @xprazak2, I did what you asked, added test and refactored the fixtures.

xprazak2 commented:
LGTM, there are test failures and style offences, but they do not look like they were caused by changes in here. I would wait with merge until they are green though.

ares commented:
[test foreman]

ezr-ondrej commented:
> LGTM, there are test failures and style offences, but they do not look like they were caused by changes in here. I would wait with merge until they are green though.

I have addressed them in #6833.

xprazak2 commented:
ok, also the react-bootstrap-typeahead will need to be removed from pacakge.json

ezr-ondrej commented:
> ok, also the react-bootstrap-typeahead will need to be removed from pacakge.json

Removed, rebased so travis is :green_heart:
Packaging is here: theforeman/foreman#3851

xprazak2 commented:
Waiting for packaging: https://github.com/theforeman/foreman-packaging/pull/3851

xprazak2 commented:
needs a rebase, waiting on https://github.com/theforeman/foreman-js/pull/34

xprazak2 commented on package.json:
```diff
@@ -72,6 +72,7 @@
     "prettier": "^1.18.0",
     "raf": "^3.4.0",
     "raw-loader": "^0.5.1",
+    "react-dnd-test-backend": "^7.4.4",
```
I think this will be pulled in via `@theforeman/vendor`, so bumping its version to 1.6.0 should make Travis happy.

Edit: Or maybe not, there is react-dnd, and react-dnd-html5-backend in foreman-js, but vendor version needs bumping regardless of that.

ezr-ondrej replied:
exactly, I do not know if the test related stuff should go in `@theforeman/vendor` or not

xprazak2 replied:
Test related things can stay here, but `@theforeman/vendor` version needs to be bumped so that we can consume react-dnd and react-dnd-html5-backend.

ezr-ondrej replied:
No need anymore, it's been bumped in other PR and even packaged (theforeman/foreman-packaging#4142)

amirfefer commented on app/helpers/form_helper.rb:
```diff
@@ -390,10 +390,17 @@ def datetime_local_f(f, attr, options = {})
     react_form_input('dateTime', f, attr, options)
   end
 
+  def orderable_select_f(f, attr, choices, select_options = {}, html_options = {})
```
can you add a test for this function?

ezr-ondrej replied:
I've added two for reordering the options - the only responsibility of this function actually.

ezr-ondrej committed:
Refs #26801 - tests for new react-dnd

ezr-ondrej commented on config/jest.config.js:
```diff
@@ -18,6 +18,11 @@ module.exports = {
+    "^react-dnd$": "react-dnd/dist/cjs/index.js",
+    "^react-dnd-test-backend$": "react-dnd-test-backend/dist/cjs/index.js",
+    "^react-dnd-test-backend/TestBackend$": "react-dnd-test-backend/dist/cjs/TestBackend.js",
+    "^react-dnd-test-utils$": "react-dnd-test-utils/dist/cjs/index.js"
```
Theese are fixing problems with jest and modules. `react-dnd` doesn't use common js implementation by default [1].

[1] http://react-dnd.github.io/react-dnd/docs/testing#setup

ezr-ondrej commented:
I've fixed tests.

amirfefer commented on webpack/assets/javascripts/react_app/components/common/forms/OrderableSelect/Orderable.stories.js:
```diff
@@ -0,0 +1,68 @@
+const StoryTag = ({ text, isDragging, value }) => {
+  const opacity = isDragging ? 0.6 : 1;
+  return (
+    <span title={value} style={Object.assign({}, style, { opacity })}>
```
instead of `Object.assign` you can use spread syntax:
```js
style={ ...style, opacity }
```

amirfefer commented on webpack/assets/javascripts/react_app/components/common/forms/OrderableSelect/OrderableSelect.js:
```diff
@@ -0,0 +1,71 @@
+  const [internalValue, setInternalValue] = useState(
+    (value || defaultValue)
+      .map(v => options.find(opt => opt.value === v))
+      .filter(v => !!v)
```
this is the state's initial value, can you move it outside this scope?

```js
const initalValue =  (value || defaultValue)
      .map(v => options.find(opt => opt.value === v))
      .filter(v => !!v)

const [internalValue, setInternalValue] = useState(initalValue);
```

ezr-ondrej replied:
I have created a custom hook for it and extracted it to a separate file entirely.

amirfefer commented on webpack/assets/javascripts/react_app/components/common/forms/OrderableSelect/OrderableSelect.js:
```diff
@@ -0,0 +1,71 @@
+  options,
+  ...props
+}) => {
+  const [internalValue, setInternalValue] = useState(
```
does it make sense to keep the value as an internal data? other components need to access it? if so, then the state should be global via redux

ezr-ondrej replied:
We keep it as an internal state, because we are not touching external value and we need to keep the order of the value somewhere. And no, other components should not need it directly, they should use `onChange` prop. We currently use `value` and `defaultValue` only as defaultValues, we can't change value through props, that's not needed in common cases though.

amirfefer commented on webpack/assets/javascripts/react_app/components/common/forms/OrderableSelect/OrderableSelect.js:
```diff
@@ -0,0 +1,71 @@
+};
+
+OrderableSelect.propTypes = {
+  options: PropTypes.arrayOf(PropTypes.object).isRequired,
```
does the object have a predefined structure? you can use `shape()`

ezr-ondrej replied:
Not really, you we need value, but the label can have diffrent key depending on the labelKey prop.

amirfefer commented on webpack/assets/javascripts/react_app/components/common/forms/OrderableSelect/helpers.js:
```diff
@@ -0,0 +1,123 @@
+import { set } from 'lodash';
+
+export const orderDragged = (inputArry, dragIndex, hoverIndex) => {
+  const dragedValue = inputArry[dragIndex];
```
typo ? `inputArray`

ezr-ondrej replied:
It wasn't - I like to avoid naming vars by type names. I agree it seems like it though. Changed to make more sense.

amirfefer commented on webpack/assets/javascripts/react_app/components/common/forms/OrderableSelect/helpers.js:
```diff
@@ -0,0 +1,123 @@
+import { DragSource, DropTarget } from 'react-dnd';
+import PropTypes from 'prop-types';
+import { set } from 'lodash';
+
```
can you add unit tests for these functions?

ezr-ondrej replied:
I have extracted the most important method and tested that one. You think it's enough? I believe it is as we have integrational test and now we test the two important helpers, only the straight forward ones are not tested.

amirfefer reviewed (changes requested):
Thanks @ezr-ondrej ! looks great :+1: 

I wonder how about using a wrapper like [this one](https://github.com/kutlugsahin/react-smooth-dnd) for `react-dnd` instead of implement helpers and test it from scratch

ezr-ondrej commented:
@amirfefer I haven't found any for the `react-dnd` as it tends to be quite simple to use. And it has far best implementation (can find an article I read when I did research on the dnd libraries to use). And according to https://www.npmtrends.com/react-dnd-vs-react-drag-list-vs-react-smooth-dnd is by far most popupar. But if you would find the wrapper for it, I am all in to use it!

amirfefer commented:
@ezr-ondrej I haven't researched drag and drop libraries yet, and as you mentioned it looks like `react-dnd` is the most promising API. but drag and drop feature is fundamental for future implementations like a new dashboard, so I just wonder if there's a wrapper that decreases the number of needed helpers functions, and makes it simpler.

there are few simple wrappers for `react-dnd` like, [this one](https://github.com/cassiozen/react-simple-dnd) or [that one](https://medium.com/@maoznir/simple-drag-and-drop-wrapper-for-react-dnd-7e7e401f80f4) - all I'm saying is to consider an interface or using an existing wrapper so it would be easier to consume and more specific to our set of needed features, also creating such an interface will allow us to use a different library in the future without changing the public interface.

EDIT: Of course this kind of ‘public interface’ can be added in a follow up PR, just some thoughts for the future :)

sharvit commented:
What is the status here?

ezr-ondrej commented:
@sharvit seems to be :green_heart: and IMHO ready.

xprazak2 commented:
just one thing I noticed - storybook creates its own components `OrderAppSandbox` and `StoryTag`, but does not show how to use `OrderableSelect` and `OrderableToken`. Is that because of some storybook limitations?

ezr-ondrej commented:
> just one thing I noticed - storybook creates its own components `OrderAppSandbox` and `StoryTag`, but does not show how to use `OrderableSelect` and `OrderableToken`. Is that because of some storybook limitations?

There is `OrderableSelect` itself in the Form component.

xprazak2 commented:
@amirfefer, any additional comments or are we good to merge?

ezr-ondrej commented:
Rebased.

amirfefer reviewed (approved):
JS stack LGTM :+1: 
Tested and works as expected
Thanks @ezr-ondrej !

ares commented:
Who has the honor to click the green button here? :-)

xprazak2 commented:
Needs approval from packaging.

tbrisker commented on package.json:
```diff
@@ -73,6 +73,8 @@
     "prettier": "^1.19.1",
     "raf": "^3.4.0",
     "raw-loader": "^0.5.1",
+    "react-dnd-test-backend": "^9.4.0",
+    "react-dnd-test-utils": "^9.4.0",
```
I'm assuming these don't need to be packaged, please add them at https://github.com/theforeman/foreman-packaging/blob/rpm/develop/update-requirements#L10-L29

ezr-ondrej replied:
Like this: theforeman/foreman-packaging#4376 ? Thanks, I didn't know it's necessary.

tbrisker replied:
it's not quite a blocker but just so the next time someone from @theforeman/packaging team syncs the foreman requirements it doesn't add this.

tbrisker reviewed (approved):
Thanks everyone!
